### DISCUSSION TITLE
with create_async_task, how to get result?

### DISCUSSION THREAD

**Discussion by fbuescher** [fhempy/fhempy#81] [Q&A]
Hi,

i want to fiddle a bit with the skodaconnect-part and wonder how the following is intended to work:

when i have the following snippet
```
async def setSomePropery(self, hash, params):
   self.create_async_task(self.setSpecialProperty(params))
```

is there a way to get a notify regarding the result of the task (succeeded or timeout or failed or exception or ...) via callback? 
I did not found any example, maybe you have short sample?

thanks,
Friedhelm

**Comment by fhempy** [comment]
Hi Friedhelm,

`self.create_async_task()` can't return a return code nor throw an exception to the "main task". The reason I often use self.create_async_task within the set_... function is that I don't want FHEM to wait for the task execution to be finished. It's a little bit like dispatching the work without waiting for the result. Nevertheless you can set FHEM readings with `fhem.readingsSingleUpdate()` or similar within that task.

Actually you might not need a callback, as you can just put the code you want to wait for within the `setSpecialProperty(params)` function.

```
async def setSomePropery(self, hash, params):
   self.create_async_task(self.setSpecialProperty(params))

async def setSpecialProperty(params):
   # do some stuff
   # you can even call blocking code here with utils.run_blocking()
   # or just execute async coroutine with await async_coroutine()
   # when everything is finished, just call the functions you would normally call as callback like
   await fhem.readingsSingleUpdate(self.hash, "state", "finished processing", 1)
```

If you still want a callback, you can try:
```
async def callback_coroutine(value):
   await fhem.readingsSingleUpdate(self.hash, "state", value, 1)

async def setSomePropery(self, hash, params):
   self.create_async_task(self.setSpecialProperty(params, self.callback_coroutine))

async def setSpecialProperty(params, callback):
   # do some stuff
   # you can even call blocking code here with utils.run_blocking()
   # or just execute async coroutine with await async_coroutine()
   # when everything is finished, just call the functions you would normally call as callback like
   await callback("finished processing")
```

all codes untested 😄 

Let me know if that explanation helps, if not let me know what you want to achieve so that I can help you in the proper context.

**Reply by fbuescher** [reply]
*many* thanks, Dominik.

In the meantime, i already tried your first suggestion. I'm now investigating why (since yesterday) my fhem dies every now and then 8-)

again, many thanks!
